jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/PullBaseImageStep.java

```diff
@@ -101,73 +88,77 @@ public ImageAndAuthorization call()
+
+    } else if (imageReference.isTagDigest()) {
+      Optional<Image> image = getCachedBaseImage();
+      if (image.isPresent()) {
+        RegistryClient noAuthRegistryClient =
+            buildContext.newBaseImageRegistryClientFactory().newRegistryClient();
+        // TODO: passing noAuthRegistryClient may be problematic. It may return 401 unauthorized if
+        // layers have to be downloaded. https://github.com/GoogleContainerTools/jib/issues/2220
+        return new ImageAndRegistryClient(image.get(), noAuthRegistryClient);
       }
     }
 
     try (ProgressEventDispatcher progressEventDispatcher =
             progressEventDispatcherFactory.create("pulling base image manifest", 2);
         TimerEventDispatcher ignored1 = new TimerEventDispatcher(eventHandlers, DESCRIPTION)) {
+
       // First, try with no credentials.
+      RegistryClient noAuthRegistryClient =
+          buildContext.newBaseImageRegistryClientFactory().newRegistryClient();
       try {
-        return new ImageAndAuthorization(pullBaseImage(null, progressEventDispatcher), null);
+        return new ImageAndRegistryClient(
+            pullBaseImage(noAuthRegistryClient, progressEventDispatcher), noAuthRegistryClient);
 
-      } catch (RegistryUnauthorizedException ignored2) {
+      } catch (RegistryUnauthorizedException ex) {
         eventHandlers.dispatch(
             LogEvent.lifecycle(
                 "The base image requires auth. Trying again for " + imageReference + "..."));
 
-        // If failed, then, retrieve base registry credentials and try with retrieved credentials.
-        // TODO: Refactor the logic in RetrieveRegistryCredentialsStep out to
-        // registry.credentials.RegistryCredentialsRetriever.
         Credential registryCredential =
-            RetrieveRegistryCredentialsStep.forBaseImage(
-                    buildContext, progressEventDispatcher.newChildProducer())
-                .call()
-                .orElse(null);
+            RegistryCredentialRetriever.getBaseImageCredential(buildContext).orElse(null);
 
-        Authorization registryAuthorization =
-            registryCredential == null || registryCredential.isOAuth2RefreshToken()
-                ? null
-                : Authorization.fromBasicCredentials(
-                    registryCredential.getUsername(), registryCredential.getPassword());
+        RegistryClient registryClient =
+            buildContext
+                .newBaseImageRegistryClientFactory()
+                .setCredential(registryCredential)
+                .newRegistryClient();
 
         try {
-          return new ImageAndAuthorization(
-              pullBaseImage(registryAuthorization, progressEventDispatcher), registryAuthorization);
+          // TODO: refactor the code (https://github.com/GoogleContainerTools/jib/pull/2202)
+          if (registryCredential == null || registryCredential.isOAuth2RefreshToken()) {
+            throw ex;
```

Comment:
This pattern is bad, but I just can't leave it to do the same meaningless no-auth attempt that is destined to fail. This is part of the issue #2202. This part should be restructured later. But I am not doing it here.

Reply:
is the restructure required before we do a 2.0.0?

Reply:
It's not required. But as of now (which doesn't have this jump), you just make the same meaningless call with null credential only to fail again. (In line 126, we already failed with "noAuthRegistryClient". No reason to do this again.) This is a simple guard, so I think this is safe. I want to have this in 2.0.0 for better performance. Most registries will hit this path.